projects/zyfai/helpers.js

```diff
@@ -0,0 +1,222 @@
+        abi: 'erc20:balanceOf',
+        calls: balanceCalls,
+    });
+    // 5. Prepare convertToAssets calls for all balances
+    const convertToAssetsCalls = balances.map((balance, i) => ({
+        target: balanceCalls[i].target,
+        params: [balance],
+    }));
+    const assets = await api.multiCall({
+        abi: 'function convertToAssets(uint256) view returns (uint256)',
+        calls: convertToAssetsCalls,
+    });
+    // 6. Normalize using asset decimals
+    let totalUsdValue = 0;
+    for (let i = 0; i < assets.length; i++) {
+        const pool = balanceCalls[i].target.toLowerCase();
+        const asset = poolToAsset[pool];
+        const decimals = decimalsMap[asset] || 18;
+        totalUsdValue += Number(assets[i]) / (10 ** decimals);
+    }
+    return totalUsdValue;
+}
+
+/**
+ * Pendle TVL
+ */
+async function pendleTvl(api, owners) {
+    // Gather all token addresses
+    const allTokens = Object.values(PENDLE_MARKET_ADDRESSES).flatMap(market => Object.values(market).map(addr => addr.toLowerCase()));
+    // Fetch prices for all tokens in one batch
+    const priceApiUrl = `https://api-v2.pendle.finance/core/v1/146/assets/prices?addresses=${allTokens.join(",")}&skip=0`;
+    const priceData = await get(priceApiUrl);
+    const prices = priceData.prices || {};
+    // Prepare all balanceOf calls for all owners and all tokens
+    const balanceCalls = allTokens.flatMap(token => owners.map(owner => ({ target: token, params: [owner] })));
+    const balances = await api.multiCall({
+        abi: 'erc20:balanceOf',
+        calls: balanceCalls,
+    });
+    // Sum USD value for all balances
+    return balances.reduce((sum, bal, i) => {
+        const token = balanceCalls[i].target.toLowerCase();
+        const price = prices[token] || 0;
+        return sum + (Number(bal) * price) / 1e18;
+    }, 0);
+}
+
+/**
+ * Beets TVL
+ */
+async function beetsTvl(api, owners) {
+    const calls = owners.flatMap(owner => [
+        { target: BEETS_POOL.deposit, params: [owner] },
+        { target: BEETS_POOL.stake, params: [owner] },
+    ]);
+    const balances = await api.multiCall({
+        abi: 'erc20:balanceOf',
+        calls,
+    });
+    const total = balances.reduce((sum, bal) => sum + Number(bal) / 1e18, 0);
+    api.add(BEETS_POOL.deposit, total);
+    return total;
+}
+
+/**
+ * Penpie TVL
+ */
+async function penpieTvl(api, owners) {
+    const balanceCalls = owners.map(owner => ({ target: PENPIE_CONTRACT, params: [PENPIE_MARKET_ADDRESS, owner] }));
+    const balances = await api.multiCall({
+        abi: 'function balance(address,address) view returns (uint256)',
+        calls: balanceCalls,
+    });
+    const priceApiUrl = `https://api-v2.pendle.finance/core/v1/146/assets/prices?addresses=${PENPIE_MARKET_ADDRESS}&skip=0`;
+    const priceData = await get(priceApiUrl);
+    const price = priceData?.prices?.[PENPIE_MARKET_ADDRESS] || 0;
+    return balances.reduce((sum, bal) => {
```

Comment:
Please leave any token pricing logic to us. Please use api.add() to export an object of {[token address]: token balance }. If any token prices are missing let us know and we will add support for them

Reply:
Okay doing it

Reply:
Hello, thanks for reviewing the PR.

I've updated the zyfai-adapter and tested it locally. I noticed that a few tokens are missing price data, which may affect the TVL calculation:

One is Beet Pool and 2 Silo Pools below,

Beets pool tokens:
'aUSDC-wstkscUSD-scUSD': {
    deposit: '0x54Ca9aad90324C022bBeD0A94b7380c03aA5884A',
    stake: '0x724a6716bf9CA384584bEb51a2eA07564c7fdD69'
}

Silo Earn pools:
Varlamore USDC Growth: 0xF6F87073cF8929C206A77b0694619DC776F89885
Re7 scUSD: 0x592D1e187729C76EfacC6dfFB9355bd7BF47B2a7

These may need price support to reflect full TVL. Let me know if anything else is needed from my side.